### ISSUE TITLE
Hide the titlebar

### ISSUE THREAD

**Issue by bl4ckb0ne** [swaywm/sway#7409]
Hiding the titlebar is done via a workaround to set the font size to 0 (see https://github.com/swaywm/sway/pull/7092 and #7408)

Font size 0 should not be allowed, instead there should be a proper way to hide the titlebars done via the config file.

Ref: 

#7350, https://github.com/swaywm/sway/issues/6946, https://github.com/swaywm/sway/issues/5050 and https://github.com/swaywm/sway/issues/1286

**Comment by Inginn** [comment]
I am at my limit, I have the 0.01 height bandaid. but I keep pressing it with my mouse if I am using chrome and jumping to another window that is behind.
I know this is not part of i3 but I could use sway befor they "fixed" it.

**Comment by WhyNotHugo** [comment]
I'd love to do this on a per-client basis. Some clients already draw functional titlebars, so I'd want to disable titlebars for those specifically.

Given that this needs to be designed from scratch, does applying it on a per-client basis sound reasonable? The all-windows use case can be covered by simply:

    for_window * titlebar hide

**Comment by rpigott** [comment]
The client can already indicate when they draw their own decorations with xdg_decoration. Only allowed for floating clients atm (otherwise we enforce ssd), though I don't remember why.

**Comment by MarcioGCosta** [comment]
Setting the default_border as pixel 0 i was able to get rid of the title bar
 
I am running version 1.8.2 in void

**Comment by rpigott** [comment]
We restored this workaround in 1.8.2 until a proper feature is added.

**Comment by hashworks** [comment]
A feature similar to `smart_gaps` would be great as well: `smart_titlebar on` would only show the titlebar when two or more windows are visible in a workspace.

**Comment by tmpm697** [comment]
> Setting the default_border as pixel 0 i was able to get rid of the title bar
> 
> I am running version 1.8.2 in void

not work for me.

currently the only way to get rid of titlebar is setting font:
```
font pango:monospace 0
```

please do not remove this setting until we can provide a solution for it, my workflow adheres with it.

**Comment by ShalokShalom** [comment]
> A feature similar to `smart_gaps` would be great as well: `smart_titlebar on` would only show the titlebar when two or more windows are visible in a workspace.

You can do that with
hide_edge_borders --i3 smart

**Comment by hashworks** [comment]
Hm, that doesn't work for me. `hide_edge_borders --i3 smart` still shows a title bar even when only one tab is open. I guess I'll open a bug report later.

**Comment by ShalokShalom** [comment]
I have it with:

```
gaps outer 0
gaps inner 0
hide_edge_borders --i3 smart
workspace_layout tabbed
```

Thats the entire config, and _it works for me_ ™️

**Comment by tmpm697** [comment]
> I have it with:
> 
> ```
> gaps outer 0
> gaps inner 0
> hide_edge_borders --i3 smart
> workspace_layout tabbed
> ```
> 
> Thats the entire config, and _it works for me_ ™️

there's still title to distinguish between tabs, I find no way to get rid of titlebar completely execpt this font hack: `font pango:monospace 0`

**Comment by Akadiuss** [comment]
> > I have it with:
> > ```
> > gaps outer 0
> > gaps inner 0
> > hide_edge_borders --i3 smart
> > workspace_layout tabbed
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Thats the entire config, and _it works for me_ tm
> 
> there's still title to distinguish between tabs, I find no way to get rid of titlebar completely execpt this font hack: `font pango:monospace 0`

With that workaround the problem is that the statue bar is affected.

**Comment by GrabbenD** [comment]
Is there any workaround which can highlight the active windows & hide the title bar?

Here's what I've tried:
- `font pango: monospace 0.001` leaves a small bar above every window
- `default_border none` doesn't highlight which window is active

**Comment by tmpm697** [comment]
> Is there any workaround which can highlight the active windows & hide the title bar?
> 
> Here's what I've tried:
> 
> * `font pango:monospace 0` shows `Invalid font size`
> * `font pango:monospace 0.001` leaves a small bar above every window
> * `default_border none` removes titlebar but doesn't highlight which window is active

try this:
```
font pango:monospace 0
default_border none
```
worked for me

**Comment by emersion** [comment]
No. This issue is about adding a new method. If this issue stays unresolved by the time 1.9 ships, there will be no more way to hide the titlebar.

**Comment by ShalokShalom** [comment]
So make this a blocker?

**Comment by emersion** [comment]
No. It's up to people who care about this feature to submit a PR.

**Comment by tmpm697** [comment]
@emersion why clean it so fast as we haven't yet come up with an PR?

**Comment by iguanajuice** [comment]
> > A feature similar to `smart_gaps` would be great as well: `smart_titlebar on` would only show the titlebar when two or more windows are visible in a workspace.
> 
> You can do that with hide_edge_borders --i3 smart

TYSM! I was always bothered by sway not behaving the same way as i3 when it comes to having no borders when using tabs, and this makes sway behave closer to i3. This really should be the default behavior, since sway is usually pretty obsessive over being 1-to-1 with i3.

**Comment by nim65s** [comment]
Hi,

I did try something in #7889. It's not exactly a proper fix for this issue, but it provide a way to hide the titlebar when there is only one window, no matter which layout is configured.

Feedback is welcome :)

**Comment by neuromagus** [comment]
Sway 1.9 released, titlebars did not disabled, hacks gone (fonts 0).

**Comment by neuromagus** [comment]
@emersion, my research has shown that changes to files ```sway/tree/container.c``` and ```sway/desktop/render.c``
are enough to completely remove headers.

IMO, need to add control structures in these files for bool option and render or not titlebar. Maybe only in render. I'm a noob and did not understand internal organization of Sway (master->child window blah-blah)...

I think it's worth paying attention here:
```
static void render_containers_linear()
static void render_containers_tabbed()
static void render_containers_stacked()
```
i tried to disable call ```render_titlebar()```, but anyway Sway render empty space for titles if I did not change in ```container.c```:
```
size_t container_titlebar_height(void) {
    //return config->font_height + config->titlebar_v_padding * 2;
    return 0;
}
```

I really hope for your correction and changes in the corrective release 1.91, @emersion.

Cheers.

**Comment by neuromagus** [comment]
well, I wrote a solution, it works. https://github.com/swaywm/sway/pull/8023#issue-2162222439

**Comment by neuromagus** [comment]
added option 'disable_titlebar' in master branch https://github.com/swaywm/sway/pull/8026#issue-2163011746

**Comment by y0nei** [comment]
Still not having a proper option to disable titlebars on windows is silly to say the least.

**Comment by neuromagus** [comment]
I'm done here. 
So, I don't know how merging works here and how long it will take for the corrected version to come out if my pull request is accepted, but I understand how titlebar in tiled WM can be annoying. Really!..

The spirit of DWM lives on, gentlemen. Here is a separate ~~hack~~ patch for Sway 1.9. ;)

Instruction:

1. Download Sway-1.9 source tarball: https://github.com/swaywm/sway/archive/refs/tags/1.9.tar.gz
2. Download patch: [disable_titlebar_patch.tar.gz](https://github.com/swaywm/sway/files/14467763/disable_titlebar_patch.tar.gz)
3. Extract Sway-1.9 tarball: ```tar xf sway-1.9.tar.gz```
4. Extract patch and move to sway-1.9 folder: ```tar xf disable_titlebar_patch.tar.gz; mv disable_titlebar.diff sway-1.9/```
5. go to folder sway-1.9 and apply the patch: ```cd sway-1.9/; patch -p1 < disable_titlebar.diff```
6. compile and install sway (Run these commands):
```
meson build/
ninja -C build/
sudo ninja -C build/ install
```
7. Edit ```~/.config/sway/config``` file and add the option: ```disable_titlebar yes``` or read ```man 5 sway``` xD
for example, a piece of my ~/.config/sway/config:
```
xwayland disable
workspace_layout tabbed
default_border none
default_floating_border none
disable_titlebar yes
```

Enjoy. That's all, folks.

**Comment by imneokitty** [comment]
> I'm done here. So, I don't know how merging works here and how long it will take for the corrected version to come out if my pull request is accepted, but I understand how titlebar in tiled WM can be annoying. Really!..
> 
> The spirit of DWM lives on, gentlemen. [Here](https://github.com/neuromagus/disable_titlebar_in_sway/tree/main) is a separate ~hack~ patch for Sway 1.9 and instruction ;)

Hello, is it possible to enable the titlebar only where there are a few windows present, but keep it disabled when there is a single one?

**Comment by imneokitty** [comment]
> I'm done here. So, I don't know how merging works here and how long it will take for the corrected version to come out if my pull request is accepted, but I understand how titlebar in tiled WM can be annoying. Really!..
> 
> The spirit of DWM lives on, gentlemen. [Here](https://github.com/neuromagus/disable_titlebar_in_sway/tree/main) is a separate ~hack~ patch for Sway 1.9 and instruction ;)

Hello, is it possible to enable the titlebar only where there are a few windows present, but keep it disabled when there is a single window?

**Comment by neuromagus** [comment]
nope. Or disable all (monocle mode/tiled) or enabled all ;} Maybe later I (or somebody) add this functional. But, for what? Titlebars eat up space.

**Comment by imneokitty** [comment]
> nope. Or disable all (monocle mode/tiled) or enabled all ;} Maybe later I (or somebody) add this functional. But, for what? Titlebars eat up space.
Whatever then. Also thought about the extra space that is being taken.
Thanks for the patch btw, works like a charm! :)

**Comment by Friptick** [comment]
The current hack (which doesn't involve a patch with a10-stage build process) is really sub-optimal: if the mouse pointer touches that `font pango:monospace 0.001` line, you will suddenly be looking a different tab. Very irritating.

As @neuromagus says
>But, for what? Titlebars eat up space.

And @kit-ty-kate says in #7350 
>I've personally never seen any sway config or screenshot where the user kept the titlebar.

But the maintainers do not seem interested, possibly for the decent-ish reason that a "tabbed layout" _without tabs_ is a bit weird.

And yet what people want here is exactly the UX you get on Android and iOS. No titlebars, just swipe (or key) left and right to get to another app (or window). On mobile nobody would imagine wasting space on titlebars, after all.

@emersion Would it not be logical therefore to fork `tabbed` into a new layout called `mobile` or `micro` or something. Ditch the titlebar there and call it a day?

**Comment by neuromagus** [comment]
@Friptick, this is not "mobile" vision... If u want learn about mobile, read this guy - https://en.wikipedia.org/wiki/Jef_Raskin ("The Humane Interface").
IMO, I just want use ALL space in my screens. I know, where is open any programs (and write layouts config). This behavior reminds DWM stile (monocle layout).
I love DWM, I very love Xmonad, but in the Wayland world we have "now" only wlroots... Wlroots -> Sway. And Sway have very big tabbars ;}

**Comment by Friptick** [comment]
OK true, you want tiling (not monocle) layout but without window chrome. And yes, tiling is not mobile UX (though 2 windows is possible on Android).

I was speculating about _why_ the maintainers think this is superfluous.

Imagine 4 windows _without titlebars_:
- tiling layout - this makes full sense, you can see what they contain already
- stacking layout - a waste of space, but as least you can see _how many_ windows there are
- tabbed layout (basically "monocle" or "mobile") - no information about what the windows contain or how many there are

Maybe that is what is discouraging the maintainers.

My point is that a mobile (monocle) layout is the ultimate space-efficient tiling layout so it is entirely appropriate in this project if it stops pretending to be "tabbed".

**Comment by neuromagus** [comment]
>My point is that a mobile (monocle) layout is the ultimate space-efficient tiling layout so it is entirely appropriate in this project if it stops pretending to be "tabbed".

ОМG... "Hot" is not "Red"!

I want option for disable "tabbar". If maintainers not implement this option, I can hack again  in the next version of Sway, like [this](https://github.com/neuromagus/disable_titlebar_in_sway). And later, we can see alternatives (DWL or River or something else) with this option. Well, True way is here (for me) - https://discourse.haskell.org/t/xmonad-for-wayland-call-for-help/7812

**Comment by Friptick** [comment]
Question seconded.

I love Sway but I'm getting fidgety waiting for the day I get my (1280x10=) **12,800 pixels** of useless blank tab real estate back, to do whatever I want with.

Thanks to anyone working on this issue, your toil will be very appreciated.

**Comment by neuromagus** [comment]
Guys, as promised before, updated Sway -> updated my patch for disable titlebar!
Here it is -> https://github.com/neuromagus/disable_titlebar_in_sway

simple instruction for install:
download Sway sources, download patch, extract all, apply patch ```patch -p1 < disable_titlebar_sway1-10.patch```, set in Sway config new option ```disable_titlebar yes``` and enjoy!

P.S. I test with borders, looks nice (i mean top border showed normal), but... If you found any bugs, because this version of Sway have too much changes -> Welcome!

**Comment by saifulapm** [comment]
> Guys, as promised before, updated Sway -> updated my patch for disable titlebar! Here it is -> https://codeberg.org/neuromagus/disable_titlebar_in_sway
> 
> simple instruction for install: download Sway sources, download patch, extract all, apply patch `patch -p1 < disable_titlebar_sway1-10.patch`, set in Sway config new option `disable_titlebar yes` and enjoy!
> 
> P.S. I test with borders, looks nice (i mean top border showed normal), but... If you found any bugs, because this version of Sway have too much changes -> Welcome!

Why not this added on Sway itself ? Why need patch ? Sway don't think this is necessary ?

**Comment by Friptick** [comment]
Thank you @neuromagus for your work.

**But unfortunately this issue cannot be considered fixed until the patch is incorporated into the project.** Personally, I am just an ordinary end user, I have hardly any experience compiling software from source. Patching and compiling involves expertise and then ongoing maintenance. It is not really that "simple".

To the maintainers here: is there any roadmap at all for incorporating this feature? Or should it be considered "won't fix"? If so maybe it's time to tag it as such?

**Comment by emersion** [comment]
It's not a wontfix, but we're still in the process of figuring out what the command API should look like. Discussion in https://github.com/swaywm/sway/pull/8248.

**Comment by neuromagus** [comment]
@Friptick. This patch works on stable version. Need compile only once. Problem only one here - @emersion takes patches to ONLY in master branch. On RELEASED version all patches uses only external.

And now community have a "process of figuring" - as i understand - "we discuss, how this feature has release and looks" maybe in titlebars, maybe in borders, or, maybe in other place...

2 lines of code ;} and discuss for discuss ;}, why not? My hack works now and next version of SwayWM released every 1 - 1.5 years... So, if u don't like titlebars and want use this WM now - use patch ;} Maybe later they will make a decision and we will get this functionality on next release, who know?

P.S. Sorry, guys, again. English is not my strong side.

**Comment by fossdd** [comment]
Thats how releases work. Once a release is stable and released it should
only contain bug fixes and patches. New changes and features should always
target the master branch.

Most users (including me) like the release model sway chose (to only
release a new version once a year). Sway kinda is critical software. If your
compositor doesn't start then you basically can't work, therefore it
shouldn't bring new features (with probably new bugs) in stable
releases. A release should be properly tested and stable.

Please target the master branch and collorborate with the developers. If
people want to use new features the can use the master branch, if they
want a stable version without new fancy features they should use the
releases.

**Comment by Piagno** [comment]
> Reminder that a new Sway release is upcoming and this feature will go away unless someone steps up.

Maybe I'm missing something, but I was able to remove the titlebar in v1.10 with: 
default_border pixel 2

**Comment by neuromagus** [comment]
@Piagno  i do not use borders, I use monocle mode, where one master window on fullscreen (60inch, 4k). If i create another window on this layout and switch - this window take a fullscreen too. Somebody uses "smartborder" - if main window alone - enabled "noborder" mode...

**Comment by Friptick** [comment]
@Piagno The hacky "solution" of `default_border pixel x|none` does not entirely remove the bar. It is still visible to the cursor so that you risk changing tabs inadvertently, which is very confusing and annoying. Adding `font pango:monospace 0` had a better effect but no longer works in the latest versions. This is all detailed above.

Thank you @emersion for your consideration of this issue.

**Comment by Juskr04** [comment]
> Guys, as promised before, updated Sway -> updated my patch for disable titlebar! Here it is -> https://codeberg.org/neuromagus/disable_titlebar_in_sway
> 
> simple instruction for install: download Sway sources, download patch, extract all, apply patch `patch -p1 < disable_titlebar_sway1-10.patch`, set in Sway config new option `disable_titlebar yes` and enjoy!
> 
> P.S. I test with borders, looks nice (i mean top border showed normal), but... If you found any bugs, because this version of Sway have too much changes -> Welcome!

When I try to build with meson build/ on the fedora sway spin, It reports an ERROR: Subproject "subprojects/wlroots" required but not found.

**Comment by neuromagus** [comment]
>Subproject "subprojects/wlroots" required but not found.
@Juskr04 
just install wlroots or wlroots-dev package and try again
